src/TypeLevel/RowList/Unordered/Indexed.purs

```diff
@@ -0,0 +1,129 @@
+module TypeLevel.RowList.Unordered.Indexed where
+
+import TypeLevel.Nat
+import Type.RowList
+import TypeLevel.RowList
+import Prim.TypeError (class Fail, Text)
+import Data.Symbol
+
+
+-- | A kind for unordered  RowLists which contain a Nat representation of their
+-- intended position in a  non-lexicographic order. Actual RowLists are automagically
+-- ordered lexicographically, Plutus Data is not ordered lexicographically, so we need
+-- this to encode the structure of Plutus Data.
+data RowListI :: forall (k :: Type). k -> Type
+data RowListI k
+
+-- Data constructors for RowListI.
+foreign import data ConsI
+  :: forall (k :: Type). Symbol -> k -> Nat  -> RowListI k -> RowListI k
+foreign import data NilI :: forall (k :: Type). RowListI k
+
+-- | Uniqueness constraint on the Nat indices of a RowListI which asserts that all indices
+-- are unique. This is needed to ensure that the various "lookup" classes return the
+-- "values" (types) that we expect.
+class UniqueIndices :: forall (k :: Type). RowListI k -> Constraint
+class UniqueIndices list
+
+instance
+  Fail (Text "Indices are not unique!") =>
+  UniqueIndices (ConsI k a n (ConsI k' a' n xs))
+else instance UniqueIndices NilI
+else instance UniqueIndices (ConsI k a n NilI)
+else instance
+  ( UniqueIndices (ConsI k a n xs)
+  , UniqueIndices (ConsI k' a' n' xs)
+  ) =>
+  UniqueIndices (ConsI k a n (ConsI k' a' n' xs))
+
+-- | Uniqueness constraint on the labels of a RowListI which asserts that all labels are unique.
+-- Again, this is needed so that the lookup functions perform in the expected manner.
+class AllUniqueLabelsI :: forall (k :: Type). RowListI k -> Constraint
+class AllUniqueLabelsI list
+
+instance
+  Fail (Text "Labels are not unique!") =>
+  AllUniqueLabelsI (ConsI k a n (ConsI k a' n' xs))
+else instance AllUniqueLabelsI NilI
+else instance AllUniqueLabelsI (ConsI k a n NilI)
+else instance
+  ( AllUniqueLabelsI (ConsI k a n xs)
+  , AllUniqueLabelsI (ConsI k' a' n' xs)
+  ) =>
+  AllUniqueLabelsI (ConsI k a n (ConsI k' a' n' xs))
+
+-- | Get the Nat index corresponding to a given symbol index in a RowListI.
+-- This is meant to be used primarily in the context of an instance declaration for some
+-- other class  (see examples in ToData.purs), and is probably not particularly useful
+-- outside of those instance contexts.
+class GetIndexWithLabel
+  :: forall (k :: Type). Symbol -> RowListI k -> Nat -> Constraint
+class IsSymbol label <= GetIndexWithLabel label list n | label list -> n
+
+instance
+  ( AllUniqueLabelsI (ConsI label a n xs)
+  , UniqueIndices (ConsI label a n xs)
+  , IsSymbol label
+  , KnownNat n
+  ) =>
+  GetIndexWithLabel label (ConsI label a n xs) n
+
+else instance
+  ( GetIndexWithLabel label xs n
+  , AllUniqueLabelsI (ConsI label' a' n' xs)
+  , UniqueIndices (ConsI label' a' n' xs)
+  , IsSymbol label
+  , KnownNat n
+  ) =>
+  GetIndexWithLabel label (ConsI label' a' n' xs) n
```

Comment:
I don't understand why this clause is needed?

Reply:
Hm, can't remember why that's there. Maybe it isn't needed?